plugins/guests/esxi/cap/public_key.rb

```diff
@@ -23,8 +23,10 @@ def self.insert_public_key(machine, contents)
           comm.execute <<-EOH.gsub(/^ {12}/, "")
             set -e
             SSH_DIR="$(grep -q '^AuthorizedKeysFile\s*\/etc\/ssh\/keys-%u\/authorized_keys$' /etc/ssh/sshd_config && echo -n /etc/ssh/keys-${USER} || echo -n ~/.ssh)"
-            mkdir -p "${SSH_DIR}"
-            chmod 0700 "${SSH_DIR}"
+            if [ ! -d "${SSH_DIR}" ]; then
```

Comment:
Since this is only checking for the existence of `$SSH_DIR`, if the dir exists but doesn't have the correct permissions the `chmod 0700 "${SSH_DIR}"` will not get run possibly resulting in unexpected behavior. So, I think this should be broken out into two checks like:
```
if [ ! -d "${SSH_DIR}" ]; then mkdir -p "${SSH_DIR}"; fi                  # Ensure dir exists
if [ $(stat -c '%a' $SSH_DIR) -lt 700 ]; chmod 0700 "${SSH_DIR}" ; fi     # Ensure dir has appropriate permissions
```

Reply:
I don't think it needs to do that, because whatever created the directory before, must do it correctly, or else that part is broken and should be fixed upstream.

Reply:
>  whatever created the directory before, must do it correctly, or else that part is broken and should be fixed upstream.

yup, I agree with you here. Though, Vagrant should probably be a bit defensive and give changing the permissions on the directory a try. Maybe something like`set +e; chmod 0700 "${SSH_DIR}"; set -e;`. In the case of an upstream issue or a user error, this will maybe remedy the situation so that the Vagrant environment is still usable. If the `chmod` command is unsuccessful like in the case of ESXI7 it will at least not error out here. However, if the `$SSH_DIR` does have bad permissions after trying to fix them (due to an upstream issue or user error) ssh-ing into the machine will fail. This failure should be dealt with upstream, or by the user.

Reply:
Will do this:

```diff
--- a/plugins/guests/esxi/cap/public_key.rb
+++ b/plugins/guests/esxi/cap/public_key.rb
@@ -23,10 +23,10 @@ module VagrantPlugins
           comm.execute <<-EOH.gsub(/^ {12}/, "")
             set -e
             SSH_DIR="$(grep -q '^AuthorizedKeysFile\s*\/etc\/ssh\/keys-%u\/authorized_keys$' /etc/ssh/sshd_config && echo -n /etc/ssh/keys-${USER} || echo -n ~/.ssh)"
-            if [ ! -d "${SSH_DIR}" ]; then
-              mkdir -p "${SSH_DIR}"
-              chmod 0700 "${SSH_DIR}"
-            fi
+            mkdir -p "${SSH_DIR}"
+            # NB in ESXi 7.0 we cannot change the /etc/ssh/keys-root directory
+            #    permissions, so ignore any errors.
+            chmod 0700 "${SSH_DIR}" || true
             cat '#{remote_path}' >> "${SSH_DIR}/authorized_keys"
             chmod 0600 "${SSH_DIR}/authorized_keys"
             rm -f '#{remote_path}'
```